libpod/runtime_img.go

```diff
@@ -256,28 +260,54 @@ func DownloadFromFile(reader *os.File) (string, error) {
 
 // LoadImage loads a container image into local storage
 func (r *Runtime) LoadImage(ctx context.Context, name, inputFile string, writer io.Writer, signaturePolicy string) (string, error) {
-	var newImages []*image.Image
-	src, err := dockerarchive.ParseReference(inputFile) // FIXME? We should add dockerarchive.NewReference()
-	if err == nil {
-		newImages, err = r.ImageRuntime().LoadFromArchiveReference(ctx, src, signaturePolicy, writer)
-	}
-	if err != nil {
-		// generate full src name with specified image:tag
-		src, err := ociarchive.NewReference(inputFile, name) // imageName may be ""
-		if err == nil {
-			newImages, err = r.ImageRuntime().LoadFromArchiveReference(ctx, src, signaturePolicy, writer)
-		}
-		if err != nil {
-			src, err := directory.NewReference(inputFile)
-			if err == nil {
-				newImages, err = r.ImageRuntime().LoadFromArchiveReference(ctx, src, signaturePolicy, writer)
+	var (
+		newImages []*image.Image
+		err       error
+		src       types.ImageReference
+	)
+
+	for _, referenceFn := range []func() (types.ImageReference, error){
+		func() (types.ImageReference, error) {
+			return dockerarchive.ParseReference(inputFile) // FIXME? We should add dockerarchive.NewReference()
+		},
+		func() (types.ImageReference, error) {
+			if src, err = ociarchive.NewReference(inputFile, name); err == nil { // imageName may be ""
+				return src, err
 			}
-			if err != nil {
-				return "", errors.Wrapf(err, "error pulling %q", name)
+			return src, err
+		},
+		func() (types.ImageReference, error) {
+			// prepend "localhost/" to support local image saved with this semantics
+			if !strings.Contains(name, "/") {
+				return ociarchive.NewReference(inputFile, fmt.Sprintf("%s/%s", image.DefaultLocalRegistry, name))
+			}
+			return src, err
```

Comment:
These `src, err` values come from some earlier iteration… It might be cleaner to return `nil, nil` and check for `src == nil` in the loop.

Same in the other copy.